fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/KeyExpressionComparisons.java

```diff
@@ -341,6 +373,96 @@ public void addToScanComparisonsBuilder(@Nonnull ScanComparisons.Builder builder
+                        if (nestedChild == null) {
+                            return null;
+                        }
+                        final List<KeyExpressionWithComparison> remainingChildren = new ArrayList<>();
+                        childrenIterator.forEachRemaining(remainingChildren::add);
+                        return nestedChild.combineUnnesting(result -> {
+                            final List<KeyExpressionWithComparison> allChildren = new ArrayList<>(children.size());
+                            allChildren.addAll(matchedChildren);
+                            allChildren.add(result);
+                            allChildren.addAll(remainingChildren);
+                            return withChildren(allChildren);
+                        });
+                    } else if (child.getMatchedComparisonType().equals(MatchedComparisonType.MATCHED)) {
+                        matchedChildren.add(child);
+                    } else {
+                        break;
+                    }
+                }
+                return null;
+            } else {
+                // The parent field is not repeated, so the view should include all parts of the then that have this
+                // nesting.
+                final List<KeyExpressionWithComparison> childrenWithMatchingNesting = new ArrayList<>();
+                final List<UnnestingFunction> unnestingFunctions = new ArrayList<>();
+                for (KeyExpressionWithComparison child : children) {
+                    final NestedAndUnnestingFunction nestedChild = child.asNestedWith(nestedContext);
+                    if (nestedChild != null) {
+                        final int nestedChildIndex = childrenWithMatchingNesting.size();
+                        childrenWithMatchingNesting.add(nestedChild.getNested());
+                        unnestingFunctions.add(result -> nestedChild.getUnnestingFunction()
+                                .unnest(result.children.get(nestedChildIndex)));
+                    } else {
+                        unnestingFunctions.add(result -> child);
+                    }
+                }
+
+                if (childrenWithMatchingNesting.isEmpty()) {
+                    return null;
+                }
+                return new NestedAndUnnestingFunction(withChildren(childrenWithMatchingNesting), result -> {
+                    List<KeyExpressionWithComparison> unnestedChildren = new ArrayList<>(result.children.size());
+                    for (int i = 0; i < children.size(); i++) {
+                        final KeyExpressionWithComparison unnestedChild = unnestingFunctions.get(i).unnest(result);
+                        if (unnestedChild == null) {
+                            return null;
+                        }
+                        unnestedChildren.add(unnestedChild);
+                    }
+                    return withChildren(unnestedChildren);
+                });
```

Comment:
Hm, how well do you think this strategy of "punching holes" out of nested fields like this works with supporting sorts. In particular, would it be smart enough to match a sort on `concat(a.b, c, a.d)` (assuming `a` is a non-repeated parent field)?